File: packages/opencode/src/util/path-validator.ts
/**
 * CLI path validation — agent feedback, not hard blocking.
 *
 * Prefer WASM (deterministic). Fall back to pure TypeScript when WASM is
 * unavailable. Used by bash tool before command execution.
 */
import path from "path"
import * as Log from "@opencode-ai/core/util/log"
import { readWasmAsset } from "./wasm-path"

const log = Log.create({ service: "util.path-validator" })

export type PathIssueCode =
  | "double_drive"
  | "system"
  | "git"
  | "outside_worktree"
  | "blocked"
  | "missing"
  | "invalid"

export type PathIssue = {
  path: string
  code: PathIssueCode
  message: string
  suggestion?: string
}

export type SandboxRules = {
  /** When false, skip validation entirely. Default true. */
  enabled?: boolean
  /** Check OS system directories. Default true. */
  system?: boolean
  /** Check .git paths. Default true. */
  git?: boolean
  /** Warn when absolute path is outside worktree. Default true. */
  outside?: boolean
  /** Warn when path does not exist (non-glob). Default true. */
  missing?: boolean
  /** Extra blocked path prefixes (case-insensitive). */
  blocked?: string[]
}

export type ValidateOptions = {
  worktree: string
  rules?: SandboxRules
}

const CODE_FROM_WASM: Record<number, PathIssueCode> = {
  1: "double_drive",
  2: "system",
  3: "git",
  4: "outside_worktree",
  5: "blocked",
  6: "invalid",
}

// Must match packages/wasm/core/src/path_validator.c (non-zero to avoid NULL).
const PATH_OFFSET = 256
const WORKTREE_OFFSET = 4352
const BLOCKED_OFFSET = 8448
const MEMORY_PAGES = 2 // 128 KiB

type WasmExports = {
  memory: WebAssembly.Memory
  pv_validate: (pathLen: number, worktreeLen: number, blockedLen: number, flags: number) => number
  pv_version: () => number
}

let _module: WebAssembly.Module | null = null
let _initPromise: Promise<WebAssembly.Module | null> | null = null
const encoder = new TextEncoder()

function rulesFlags(rules?: SandboxRules): number {
  let flags = 0
  if (rules?.system !== false) flags |= 1
  if (rules?.git !== false) flags |= 2
  if (rules?.outside !== false) flags |= 4
  return flags
}

function normalizeSeps(p: string): string {
  return p.replace(/\\/g, "/")
}

function isAbsolutePath(p: string): boolean {
  if (p.startsWith("/") || p.startsWith("\\")) return true
  return /^[A-Za-z]:[\\/]/.test(p)
}

function fixDoubleDrive(p: string): string | undefined {
  const m = p.match(/^([A-Za-z]:)[\\/]\1(.*)$/i)
  if (!m) return undefined
  return m[1] + (p.includes("\\") ? "\\" : "/") + (m[2] || "").replace(/^[/\\]+/, "")
}

function issueForCode(p: string, code: PathIssueCode): PathIssue {
  switch (code) {
    case "double_drive": {
      const fix = fixDoubleDrive(p)
      return {
        path: p,
        code,
        message: `"${p}" — invalid: double drive letter`,
        suggestion: fix ? `use ${fix}` : undefined,
      }
    }
    case "system":
      return { path: p, code, message: `"${p}" — blocked: system directory` }
    case "git":
      return { path: p, code, message: `"${p}" — blocked: .git directory` }
    case "outside_worktree":
      return {
        path: p,
        code,
        message: `"${p}" — outside worktree (use external_directory permission or navigation.allow)`,
      }
    case "blocked":
      return { path: p, code, message: `"${p}" — blocked by sandbox.rules` }
    case "invalid":
      return { path: p, code, message: `"${p}" — invalid path` }
    case "missing":
      return { path: p, code, message: `"${p}" — path does not exist` }
  }
}

/** Pure TypeScript validation (fallback + parity with WASM). */
export function validatePathTs(p: string, opts: ValidateOptions): PathIssue | null {
  const rules = opts.rules
  if (rules?.enabled === false) return null
  if (!p) return issueForCode(p, "invalid")

  if (/^[A-Za-z]:[\\/][A-Za-z]:/.test(p)) return issueForCode(p, "double_drive")

  if (rules?.system !== false) {
    if (/^(C:\\Windows|C:\/Windows|\/etc|\/usr|\/bin|\/sbin|\/var|\/root|\/System|\/Library)([\\/]|$)/i.test(p)) {
      return issueForCode(p, "system")
    }
  }

  if (rules?.git !== false) {
    if (/(^|[\\/])\.git([\\/]|$)/.test(p)) return issueForCode(p, "git")
  }

  const blocked = rules?.blocked ?? []
  for (const prefix of blocked) {
    if (!prefix) continue
    const a = normalizeSeps(p).toLowerCase()
    const b = normalizeSeps(prefix).toLowerCase().replace(/\/+$/, "")
    if (a === b || a.startsWith(b + "/")) return issueForCode(p, "blocked")
  }

  if (rules?.outside !== false && isAbsolutePath(p) && opts.worktree) {
    const abs = path.normalize(p)
    const wt = path.normalize(opts.worktree)
    const rel = path.relative(wt, abs)
    if (rel.startsWith("..") || path.isAbsolute(rel)) {
      return issueForCode(p, "outside_worktree")
    }
  }

  return null
}

async function loadModule(): Promise<WebAssembly.Module | null> {
  if (_module) return _module
  if (_initPromise) return _initPromise
  _initPromise = (async () => {
    try {
      const asset = await readWasmAsset("path_validator.wasm")
      if (!asset.bytes) {
        log.debug("path-validator: WASM not found, using TS fallback", { tried: asset.tried })
        return null
      }
      _module = await WebAssembly.compile(asset.bytes)
      log.info("path-validator: WASM loaded from " + asset.path)
      return _module
    } catch (err) {
      log.warn("path-validator: WASM load failed", { error: String(err) })
      return null
    }
  })()
  return _initPromise
}

export async function initPathValidator(): Promise<boolean> {
  return (await loadModule()) !== null
}

function writeUtf8(mem: WebAssembly.Memory, offset: number, text: string, max: number): number {
  const bytes = encoder.encode(text)
  const len = Math.min(bytes.length, max)
  new Uint8Array(mem.buffer, offset, len).set(bytes.subarray(0, len))
  return len
}

function writeBlocked(mem: WebAssembly.Memory, prefixes: string[]): number {
  const view = new Uint8Array(mem.buffer)
  let off = BLOCKED_OFFSET
  const end = BLOCKED_OFFSET + 4096 - 2
  for (const p of prefixes) {
    const b = encoder.encode(p)
    if (off + b.length + 2 > end) break
    view.set(b, off)
    off += b.length
    view[off++] = 0
  }
  view[off++] = 0
  return off - BLOCKED_OFFSET
}

async function validateOneWasm(p: string, opts: ValidateOptions): Promise<PathIssue | null | undefined> {
  const mod = await loadModule()
  if (!mod) return undefined

  try {
    const memory = new WebAssembly.Memory({ initial: MEMORY_PAGES, maximum: MEMORY_PAGES })
    const instance = await WebAssembly.instantiate(mod, {
      env: { memory },
    })
    const exp = instance.exports as unknown as WasmExports
    if (typeof exp.pv_validate !== "function") return undefined

    const pathLen = writeUtf8(memory, PATH_OFFSET, p, 4095)
    const wtLen = writeUtf8(memory, WORKTREE_OFFSET, opts.worktree ?? "", 4095)
    const blocked = opts.rules?.blocked ?? []
    const blockedLen = writeBlocked(memory, blocked)
    const code = exp.pv_validate(pathLen, wtLen, blockedLen, rulesFlags(opts.rules)) >>> 0
    if (code === 0) return null
    const mapped = CODE_FROM_WASM[code]
    if (!mapped) return null
    return issueForCode(p, mapped)
  } catch (err) {
    log.debug("path-validator: WASM validate failed", { error: String(err), path: p })
    return undefined
  }
}

/**
 * Validate paths. Uses WASM when available; otherwise TS.
 * Also appends existence checks in TS (needs host fs).
 */
export async function validatePaths(paths: string[], opts: ValidateOptions): Promise<PathIssue[]> {
  if (opts.rules?.enabled === false) return []
  const issues: PathIssue[] = []
  const seen = new Set<string>()

  for (const p of paths) {
    if (!p || seen.has(p)) continue
    seen.add(p)

    let issue = await validateOneWasm(p, opts)
    if (issue === undefined) {
      issue = validatePathTs(p, opts)
    }
    if (issue) {
      issues.push(issue)
      continue
    }

    // Existence is host-only (WASM has no FS).
    if (opts.rules?.missing !== false && !p.includes("*") && !p.includes("?")) {
      try {
        const { existsSync } = await import("fs")
        const resolved = path.isAbsolute(p) ? p : path.resolve(opts.worktree, p)
        if (!existsSync(resolved)) {
          issues.push(issueForCode(p, "missing"))
        }
      } catch (err) {
        log.debug("path-validator: existence check failed", { path: p, error: String(err) })
      }
    }
  }

  return issues
}

export function formatPathIssues(issues: PathIssue[]): string | undefined {
  if (issues.length === 0) return undefined
  const lines = issues.map((issue, i) => {
    const base = `  ${i + 1}. ${issue.message}`
    return issue.suggestion ? `${base}\n     Suggested fix: ${issue.suggestion}` : base
  })
  return `⚠ Path issues detected:\n${lines.join("\n")}\n\nThese are warnings — the command still ran. Fix paths if the command failed.`
}

export * as PathValidator from "./path-validator"

File: packages/wasm/core/src/path_validator.c
/*
 * Path validator for CLI commands — compiled to WASM.
 *
 * Deterministic, no-stdlib checks used before shell execution to produce
 * agent-facing feedback (not hard blocks). The host TS layer formats the report.
 *
 * Memory (imported from host):
 *   Host writes UTF-8 path at PATH_OFFSET, worktree at WORKTREE_OFFSET,
 *   optional blocked-prefix list at BLOCKED_OFFSET (NUL-separated, double-NUL end).
 *
 * Exports:
 *   pv_validate(path_len, worktree_len, blocked_len, flags) -> issue code
 *     0 = ok
 *     1 = double drive letter (e.g. D:\D:\path)
 *     2 = system directory
 *     3 = .git path
 *     4 = outside worktree
 *     5 = blocked prefix (config)
 *     6 = empty / invalid path
 *
 * flags bit0 = check system, bit1 = check git, bit2 = check outside worktree
 */

typedef __SIZE_TYPE__ size_t;
typedef __UINT8_TYPE__ uint8_t;
typedef __UINT32_TYPE__ uint32_t;
#define NULL ((void *)0)

/* Offsets must be non-zero: C treats pointer value 0 as NULL and optimizes away loads. */
#define PATH_OFFSET 256
#define WORKTREE_OFFSET 4352
#define BLOCKED_OFFSET 8448
#define MAX_PATH 4096
/* Host provides linear memory via env.memory import; we only read fixed offsets. */

static int is_sep(unsigned char c) {
  return c == '/' || c == '\\';
}

static int is_alpha(unsigned char c) {
  return (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z');
}

static unsigned char lower(unsigned char c) {
  if (c >= 'A' && c <= 'Z') return (unsigned char)(c + 32);
  return c;
}

static int ieq_n(const unsigned char *a, const unsigned char *b, size_t n) {
  size_t i;
  for (i = 0; i < n; i++) {
    if (lower(a[i]) != lower(b[i])) return 0;
  }
  return 1;
}

/* Match prefix then end-of-string or path separator. */
static int prefix_boundary_ci(const unsigned char *s, size_t slen, const char *pfx) {
  size_t i = 0;
  while (pfx[i]) {
    if (i >= slen) return 0;
    if (lower(s[i]) != lower((unsigned char)pfx[i])) return 0;
    i++;
  }
  if (i == slen) return 1;
  return is_sep(s[i]);
}

static int contains_git(const unsigned char *s, size_t len) {
  size_t i;
  for (i = 0; i + 4 <= len; i++) {
    if ((i == 0 || is_sep(s[i - 1])) &&
        lower(s[i]) == '.' &&
        lower(s[i + 1]) == 'g' &&
        lower(s[i + 2]) == 'i' &&
        lower(s[i + 3]) == 't' &&
        (i + 4 == len || is_sep(s[i + 4]))) {
      return 1;
    }
  }
  return 0;
}

static int is_system(const unsigned char *s, size_t len) {
  static const char *sys[] = {
      "C:\\Windows",
      "C:/Windows",
      "/etc",
      "/usr",
      "/bin",
      "/sbin",
      "/var",
      "/root",
      "/System",
      "/Library",
      NULL,
  };
  size_t i;
  for (i = 0; sys[i]; i++) {
    if (prefix_boundary_ci(s, len, sys[i])) return 1;
  }
  return 0;
}

/* Double drive: X:\Y: or X:/Y: */
static int is_double_drive(const unsigned char *s, size_t len) {
  if (len < 5) return 0;
  if (!is_alpha(s[0]) || s[1] != ':') return 0;
  if (!is_sep(s[2])) return 0;
  if (!is_alpha(s[3]) || s[4] != ':') return 0;
  return 1;
}

/* Outside worktree: path absolute and not under worktree (case-insensitive, sep-normalized). */
static int is_outside_worktree(const unsigned char *path, size_t plen,
                               const unsigned char *wt, size_t wlen) {
  size_t i;
  if (wlen == 0 || plen == 0) return 0;
  /* Relative paths are considered inside (resolved by host). */
  int absolute = 0;
  if (is_sep(path[0])) absolute = 1;
  else if (plen >= 2 && is_alpha(path[0]) && path[1] == ':') absolute = 1;
  if (!absolute) return 0;

  /* Strip trailing seps on worktree */
  while (wlen > 0 && is_sep(wt[wlen - 1])) wlen--;
  if (wlen == 0) return 0;

  if (plen < wlen) return 1;
  for (i = 0; i < wlen; i++) {
    unsigned char a = path[i];
    unsigned char b = wt[i];
    if (is_sep(a) && is_sep(b)) continue;
    if (lower(a) != lower(b)) return 1;
  }
  if (plen == wlen) return 0;
  return !is_sep(path[wlen]);
}

static int is_blocked(const unsigned char *path, size_t plen,
                      const unsigned char *blocked, size_t blen) {
  size_t i = 0;
  if (blen == 0 || !blocked) return 0;
  while (i < blen) {
    size_t start = i;
    size_t len = 0;
    if (blocked[i] == 0) break;
    while (i < blen && blocked[i] != 0) {
      len++;
      i++;
    }
    if (len > 0) {
      /* Manual prefix_boundary with non-NUL-terminated prefix */
      if (plen >= len && ieq_n(path, blocked + start, len) &&
          (plen == len || is_sep(path[len]))) {
        return 1;
      }
    }
    if (i < blen && blocked[i] == 0) i++; /* skip NUL */
  }
  return 0;
}

__attribute__((export_name("pv_validate")))
uint32_t pv_validate(uint32_t path_len, uint32_t worktree_len, uint32_t blocked_len, uint32_t flags) {
  const unsigned char *path = (const unsigned char *)(size_t)PATH_OFFSET;
  const unsigned char *wt = (const unsigned char *)(size_t)WORKTREE_OFFSET;
  const unsigned char *blocked = (const unsigned char *)(size_t)BLOCKED_OFFSET;

  if (path_len == 0 || path_len > MAX_PATH) return 6;
  if (worktree_len > MAX_PATH) worktree_len = MAX_PATH;
  if (blocked_len > MAX_PATH) blocked_len = MAX_PATH;

  if (is_double_drive(path, path_len)) return 1;

  if ((flags & 1u) && is_system(path, path_len)) return 2;
  if ((flags & 2u) && contains_git(path, path_len)) return 3;
  if ((flags & 4u) && is_outside_worktree(path, path_len, wt, worktree_len)) return 4;
  if (is_blocked(path, path_len, blocked, blocked_len)) return 5;

  return 0;
}

/* Keep a tiny dummy so --export-dynamic has something if needed. */
__attribute__((export_name("pv_version")))
uint32_t pv_version(void) {
  return 1;
}
